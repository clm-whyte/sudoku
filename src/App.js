import React from "react";
import { Provider } from "react-redux";
import store from "./lib/redux";

import Cell from "./components/Cell";

import "./index.scss";
function App() {
  return (
    <Provider store={store}>
      <Cell />
    </Provider>
  );
}
export default App;
