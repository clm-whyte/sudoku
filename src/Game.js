import React from "react";

import Grid from "./components/Grid";

import "./index.scss";
class Game extends React.Component {
  render() {
    return (
      <div className="game">
        <div className="grid">
          <Grid />
        </div>
      </div>
    );
  }
}
export default Game;
