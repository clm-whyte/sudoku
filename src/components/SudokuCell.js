import React from "react";
import Button from "@material-ui/core/Button";
import { StylesProvider } from "@material-ui/core/styles";

export default function SudokuCell(props) {
  return (
    <StylesProvider injectFirst>
      <Button
        variant="outlined"
        aria-label={props.cellID}
        data-test={props.cellID}
        onClick={props.onClick}
        selected={props.selected}
        className={`sudokuCell
        ${props.row % 3 === 0 ? "boxBoundaryTop" : ""}
        ${props.row % 3 === 2 ? "boxBoundaryBottom" : ""}
        ${props.row === 8 ? "boxBoundaryBottom" : ""}
        ${props.col % 3 === 0 ? "boxBoundaryLeft" : ""}
        ${props.col % 3 === 2 ? "boxBoundaryRight" : ""}
        ${props.col === 8 ? "boxBoundaryRight" : ""}
        ${props.selected ? "selected" : ""}
        `}
      >
        {props.value}
      </Button>
    </StylesProvider>
  );
}
