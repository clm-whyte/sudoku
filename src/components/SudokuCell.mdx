import { Story, Preview } from "@storybook/addon-docs/blocks";
import SudokuCell from "./SudokuCell.js";

# Cell

In standard Sudoku, the grid is made up of 81 **cells** in a 9x9 square (9 rows and 9 columns).  
The **cells** are arranged into 3x3 sections called **boxes**, which are outlined in bold.

## Default

By default, a cell is has no value, is unselected and has a thin grey outline.  
The cells are **Material UI** Buttons which have been restyled using some Sassy CSS.

```css
.sudokuCell {
  max-width: 34px;
  max-height: $cellSize;
  min-width: 34px;
  min-height: $cellSize;
  border: $cellBorder;
  border-radius: 0;
  background-color: white;
  color: navy;
}
```

<Preview>
  <Story id="cell--empty" />
</Preview>

## Position

A cell's position in the grid determines its outline.  
If it's in the first row/column, it will have a thick top/left border respectively.  
Otherwise, every 3rd row/column is given a thick bottom/right border, based on the following lines of code:

```javascript
${props.col % 3 === 2 ? "boxBoundaryRight" : ""}
${props.row % 3 === 2 ? "boxBoundaryBottom" : ""}
```

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--outer-corner" />
  <Story id="cell--inner-corner" />
</Preview>

## Selection

By default a cell is unselected.  
An unselected cell has a white background and may or may not be filled with a value.

A selected cell is denoted by a pink background colour.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--selected" />
</Preview>

## Cursor

There can be only one "cursor" Cell on the grid.  
The cursor shows the current position of the "active" cell. By default, this is at position `0, 0` and unselected.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--cursor" />
  <Story id="cell--cursor-and-selected" />
</Preview>

## Value

The cell values default to a value of `null`.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--filled" />
</Preview>
