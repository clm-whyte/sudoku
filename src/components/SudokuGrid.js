import React from "react";
import SudokuCell from "./SudokuCell";
import Grid from "@material-ui/core/Grid";

class SudokuGrid extends React.Component {
  constructor(props) {
    super(props);
    const grid = [];
    for (let i = 0; i < 9; i++) {
      grid.push(Array(9).fill({ value: null, selected: false }));
    }
    this.state = { cells: grid };
  }

  deselectCell(index) {
    index.selected = false;
  }

  deselectAllCells() {
    const cells = this.state.cells.slice();
    for (let row = 0; row < cells[row].length; row++) {
      cells[row].forEach(this.deselectCell);
    }
    this.setState({ cells });
  }

  handleClick(row, col) {
    const cells = this.state.cells.slice();
    cells[row][col] = { value: cells[row].length * row + col, selected: true };
    this.setState({ cells });
  }

  renderCell(row, col) {
    return (
      <SudokuCell
        value={this.state.cells[row][col].value}
        selected={this.state.cells[row][col].selected}
        key={this.state.cells[row].length * row + col}
        cellID={`cell-${row}/${col}`}
        onClick={() => this.handleClick(row, col)}
        row={row}
        col={col}
      />
    );
  }

  render() {
    let grid = [];
    for (let rowNumber = 0; rowNumber < 9; rowNumber++) {
      let row = [];
      for (let colNumber = 0; colNumber < 9; colNumber++) {
        row.push(this.renderCell(rowNumber, colNumber));
      }
      grid.push(
        <div
          data-test={`row-${rowNumber}`}
          key={"row" + rowNumber}
          className="grid-row"
        >
          {row}
        </div>
      );
    }
    return (
      <Grid className="grid" data-test="grid">
        {grid}
      </Grid>
    );
  }
}

export default SudokuGrid;
