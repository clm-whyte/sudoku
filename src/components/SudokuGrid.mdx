import { Story, Preview } from "@storybook/addon-docs/blocks";
import SudokuGrid from "./SudokuGrid.js";

# Grid

I'm going to try and add a little bit about the default Cell here.

<Story id="grid--empty" />

<Preview>
  <Story id="grid--empty"></Story>
</Preview>
