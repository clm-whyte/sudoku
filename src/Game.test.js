/// <reference types="jest" />

import React from "react";

import { shallow } from "enzyme";

import App from "./App";
import Game from "./Game";
import SudokuGrid from "./components/SudokuGrid";
import SudokuCell from "./components/SudokuCell";

test("verify that app contains game", () => {
  const app = shallow(<App />);
  const game = <Game />;
  expect(app).toContainReact(game);
});

test("verify that game contains grid", () => {
  const game = shallow(<Game />);
  const grid = <SudokuGrid />;
  expect(game).toContainReact(grid);
});

test("test shallow render of Sudoku Grid", () => {
  const grid = [];
  for (let i = 0; i < 9; i++) {
    grid.push(Array(9).fill({ value: null, selected: false }));
  }

  const sudokuGrid = shallow(<SudokuGrid state={grid} />);
  expect(sudokuGrid).toMatchSnapshot();
  sudokuGrid.unmount();
});

test("test state change of Sudoku Grid", () => {
  const empty = [];
  for (let i = 0; i < 9; i++) {
    empty.push(Array(9).fill({ value: null, selected: false }));
  }

  const selected = empty.slice();
  selected[0] = selected[0].map((cell, index) => {
    if (index === 0) {
      return (cell = { value: 0, selected: true });
    }
    return cell;
  });

  const sudokuGrid = shallow(<SudokuGrid state={empty} />);
  const cell = sudokuGrid.find(SudokuCell).at(0);

  expect(sudokuGrid.state().cells).toEqual(empty);
  cell.simulate("click");
  expect(sudokuGrid.state().cells).toEqual(selected);
  sudokuGrid.instance().deselectAllCells();
  expect(sudokuGrid.state().cells).toEqual(empty);
});
