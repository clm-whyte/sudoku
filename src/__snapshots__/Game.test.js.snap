// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test shallow render of Sudoku Grid 1`] = `
<WithStyles(ForwardRef(Grid))
  className="grid"
  data-test="grid"
>
  <div
    className="grid-row"
    data-test="row-0"
    key="row0"
  >
    <SudokuCell
      cellID="cell-0/0"
      col={0}
      key="0"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/1"
      col={1}
      key="1"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/2"
      col={2}
      key="2"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/3"
      col={3}
      key="3"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/4"
      col={4}
      key="4"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/5"
      col={5}
      key="5"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/6"
      col={6}
      key="6"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/7"
      col={7}
      key="7"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-0/8"
      col={8}
      key="8"
      onClick={[Function]}
      row={0}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-1"
    key="row1"
  >
    <SudokuCell
      cellID="cell-1/0"
      col={0}
      key="9"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/1"
      col={1}
      key="10"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/2"
      col={2}
      key="11"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/3"
      col={3}
      key="12"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/4"
      col={4}
      key="13"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/5"
      col={5}
      key="14"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/6"
      col={6}
      key="15"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/7"
      col={7}
      key="16"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-1/8"
      col={8}
      key="17"
      onClick={[Function]}
      row={1}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-2"
    key="row2"
  >
    <SudokuCell
      cellID="cell-2/0"
      col={0}
      key="18"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/1"
      col={1}
      key="19"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/2"
      col={2}
      key="20"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/3"
      col={3}
      key="21"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/4"
      col={4}
      key="22"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/5"
      col={5}
      key="23"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/6"
      col={6}
      key="24"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/7"
      col={7}
      key="25"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-2/8"
      col={8}
      key="26"
      onClick={[Function]}
      row={2}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-3"
    key="row3"
  >
    <SudokuCell
      cellID="cell-3/0"
      col={0}
      key="27"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/1"
      col={1}
      key="28"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/2"
      col={2}
      key="29"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/3"
      col={3}
      key="30"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/4"
      col={4}
      key="31"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/5"
      col={5}
      key="32"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/6"
      col={6}
      key="33"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/7"
      col={7}
      key="34"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-3/8"
      col={8}
      key="35"
      onClick={[Function]}
      row={3}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-4"
    key="row4"
  >
    <SudokuCell
      cellID="cell-4/0"
      col={0}
      key="36"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/1"
      col={1}
      key="37"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/2"
      col={2}
      key="38"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/3"
      col={3}
      key="39"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/4"
      col={4}
      key="40"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/5"
      col={5}
      key="41"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/6"
      col={6}
      key="42"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/7"
      col={7}
      key="43"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-4/8"
      col={8}
      key="44"
      onClick={[Function]}
      row={4}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-5"
    key="row5"
  >
    <SudokuCell
      cellID="cell-5/0"
      col={0}
      key="45"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/1"
      col={1}
      key="46"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/2"
      col={2}
      key="47"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/3"
      col={3}
      key="48"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/4"
      col={4}
      key="49"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/5"
      col={5}
      key="50"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/6"
      col={6}
      key="51"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/7"
      col={7}
      key="52"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-5/8"
      col={8}
      key="53"
      onClick={[Function]}
      row={5}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-6"
    key="row6"
  >
    <SudokuCell
      cellID="cell-6/0"
      col={0}
      key="54"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/1"
      col={1}
      key="55"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/2"
      col={2}
      key="56"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/3"
      col={3}
      key="57"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/4"
      col={4}
      key="58"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/5"
      col={5}
      key="59"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/6"
      col={6}
      key="60"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/7"
      col={7}
      key="61"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-6/8"
      col={8}
      key="62"
      onClick={[Function]}
      row={6}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-7"
    key="row7"
  >
    <SudokuCell
      cellID="cell-7/0"
      col={0}
      key="63"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/1"
      col={1}
      key="64"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/2"
      col={2}
      key="65"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/3"
      col={3}
      key="66"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/4"
      col={4}
      key="67"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/5"
      col={5}
      key="68"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/6"
      col={6}
      key="69"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/7"
      col={7}
      key="70"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-7/8"
      col={8}
      key="71"
      onClick={[Function]}
      row={7}
      selected={false}
      value={null}
    />
  </div>
  <div
    className="grid-row"
    data-test="row-8"
    key="row8"
  >
    <SudokuCell
      cellID="cell-8/0"
      col={0}
      key="72"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/1"
      col={1}
      key="73"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/2"
      col={2}
      key="74"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/3"
      col={3}
      key="75"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/4"
      col={4}
      key="76"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/5"
      col={5}
      key="77"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/6"
      col={6}
      key="78"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/7"
      col={7}
      key="79"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
    <SudokuCell
      cellID="cell-8/8"
      col={8}
      key="80"
      onClick={[Function]}
      row={8}
      selected={false}
      value={null}
    />
  </div>
</WithStyles(ForwardRef(Grid))>
`;
